src/overlay/Peer.cpp

```diff
@@ -1840,6 +1964,32 @@ Peer::recvSurveyResponseMessage(StellarMessage const& msg)
         msg, shared_from_this());
 }
 
+void
+Peer::queueTxHashAndMaybeTrim(Hash const& hash)
+{
+    mIncomingTxHashes.push_back(hash);
+    if (mIncomingTxHashes.size() > mApp.getHerder().getMaxQueueSizeOps())
+    {
+        mIncomingTxHashes.pop_front();
+    }
+}
+
+void
+Peer::recvFloodAdvert(StellarMessage const& msg)
+{
+    for (Hash const& hash : msg.floodAdvert().txHashes)
```

Comment:
should we only process the first `N` (`FLOOD_MAX_ADVERT_SIZE`?) items?

Reply:
I feel that `FLOOD_MAX_ADVERT_SIZE` might be too aggressive. For instance, if Node 1 sets it to 50 and Node 2 sets it to 100, half of the tx hashes from Node 2 will always be ignored by Node 1.

I instead changed it to process up to `getMaxQueueSizeOps` as that's the threshold that `TxDemandQueue` uses.

Reply:
this is incorrect: you should let the queueAndMaybeTrim do the trimming (and separate responsibilities properly). Right now if there are getLastMaxTxSetSizeOps + 1 hashes incoming, you'd drop the first getLastMaxTxSetSizeOps, whereas what should happen is that the first getLastMaxTxSetSizeOps will push the older hashes out of the queue in queueAndMaybeTrim. That being said, queueAndMaybeTrim should probably take a range as opposed to individual message, so you can do something more performant there than iterating over each individual hash.

Reply:
I changed the argument of `queueAndMaybeTrim` to a vector instead of an element.

Reply:
@marta-lokhova 

I'm not sure I understand. I wanted the `if` statement to take care of the case when `txHashes` has more than `getLastMaxTxSetSizeOps` hashes and thus it kicks out everything in the queue. 

For instance, `getLastMaxTxSetSizeOps = 50` and `txHashes` has 60 hashes (e.g., `{hash 1, hash 2, ..., hash 60}`). Then `TxAdvertQueue` should have the last 50 hashes in the `txHashes`. (`{hash 11, hash 12, ..., hash 60}`). I wanted the `if` statement to avoid adding `{hash 1, hash 2, ..., hash 10}`.


This was my best idea to follow the original suggestion:

> should we only process the first N (FLOOD_MAX_ADVERT_SIZE?) items?"

Although, now that I'm looking back at this, this looks like a micro optimization without much value. Your approach should work correctly but it requires multiple cases as `TxAdvertQueue` consists of two queues. I'm not sure it'll be noticeably faster since `erase` is a linear operation. I don't think the current "optimization" in the PR is really worth the extra logic either.

I think that we should just simply add all elements and pop until we hit the limit. I don't think reducing a possible number of `pop` and `push_back` is worth it. (Also, I just remembered that this original suggestion was made when the XDR type wasn't bounded, but it is now.)

Let me know what you think!

Reply:
Updated for clarity based on the discussion offline.